systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+}
+
+// StartNewCollector creates a new collector and starts
+// querying expvar endpoint at the specified interval.
+func StartNewCollector(ctx context.Context, serverURL string, period time.Duration) (*Collector, error) {
+	c := &Collector{
+		serverURL: serverURL,
+		period:    period,
+		metrics:   make(map[Metric]*AggregateStats),
+		watchers:  make(map[Metric]*watchItem),
+	}
+	return c, c.start(ctx)
+}
+
+// Get returns the aggregated stat of a given metric.
+func (c *Collector) Get(m Metric) AggregateStats {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	stats := c.metrics[m]
+	statsCopy := *stats
+	return statsCopy
+}
+
+// Delta returns the diff of a metric valuefrom the first
+// event that was collected by the collector.
+func (c *Collector) Delta(m Metric) int64 {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	stats := c.metrics[m]
+	return stats.Last - stats.First
+}
+
+// AddWatch configures a new watcher for a given metric. The
+// watcher is deleted after the specified expected value is
+// observed.
+//
+// Watcher returns a read only channel to observe the state
+// of the watch. A true event on this channel refers to an
+// observation with the expected value while a false event
+// refers to an error in the collector. Both cases end the
+// lifecycle of a watcher.
+func (c *Collector) AddWatch(m Metric, expected int64) (<-chan bool, error) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if _, ok := c.watchers[m]; ok {
+		return nil, errors.New("watcher already exists for the given metric")
+	}
+
+	out := make(chan bool)
+	c.watchers[m] = &watchItem{
+		notifyChan: out,
+		validator:  func(in int64) bool { return in == expected },
+	}
+
+	return out, nil
+}
+
+func (c *Collector) addExpvar(e expvar) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.processMetric(Goroutines, e.Goroutines)
+	c.processMetric(Bytes, e.UncompressedBytes)
+	c.processMetric(TotalEvents, e.TotalEvents)
+	c.processMetric(ActiveEvents, e.ActiveEvents)
+	c.processMetric(RSSMemoryBytes, e.RSSMemoryBytes)
+	c.processMetric(AvailableBulkRequests, e.AvailableBulkRequests)
+	c.processMetric(TransactionsProcessed, e.TransactionsProcessed)
+	c.processMetric(SpansProcessed, e.SpansProcessed)
+	c.processMetric(MetricsProcessed, e.MetricsProcessed)
+	c.processMetric(ErrorsProcessed, e.ErrorsProcessed)
+	c.processMetric(ErrorElasticResponses, e.ErrorElasticResponses)
+	c.processMetric(ErrorOTLPTracesResponses, e.ErrorOTLPTracesResponses)
+	c.processMetric(ErrorOTLPMetricsResponses, e.ErrorOTLPMetricsResponses)
+	c.processMetric(NumGC, int64(e.NumGC))
+	c.processMetric(PauseTotalNs, int64(e.PauseTotalNs))
+	c.processMetric(MemAllocs, int64(e.Mallocs))
+	c.processMetric(MemBytes, int64(e.TotalAlloc))
+	c.processMetric(HeapAlloc, int64(e.HeapAlloc))
+	c.processMetric(HeapObjects, int64(e.HeapObjects))
+}
+
+func (c *Collector) processMetric(m Metric, val int64) {
+	if _, ok := c.metrics[m]; !ok {
+		c.metrics[m] = &AggregateStats{}
+	}
+	c.updateMetric(c.metrics[m], val)
```

Comment:
What do you think about changing this to not use the pointer approach? Looks slightly more idiomatic and easier to parse:

```suggestion
	stats := c.metrics[m]
	c.metrics[m] = c.updateMetric(stats, val)
```